flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriteRequestExecutorImpl.java

```diff
@@ -0,0 +1,161 @@
+
+	private final ChannelStateWriteRequestDispatcher dispatcher;
+	private final BlockingDeque<ChannelStateWriteRequest> deque;
+	private final Thread thread;
+	private volatile Exception thrown = null;
+	private volatile boolean wasClosed = false;
+
+	ChannelStateWriteRequestExecutorImpl(ChannelStateWriteRequestDispatcher dispatcher) {
+		this(dispatcher, new LinkedBlockingDeque<>(DEFAULT_HANDOVER_CAPACITY));
+	}
+
+	ChannelStateWriteRequestExecutorImpl(ChannelStateWriteRequestDispatcher dispatcher, BlockingDeque<ChannelStateWriteRequest> deque) {
+		this.dispatcher = dispatcher;
+		this.deque = deque;
+		this.thread = new Thread(this::run);
+		this.thread.setDaemon(true);
+	}
+
+	@VisibleForTesting
+	void run() {
+		try {
+			loop();
+		} catch (Exception ex) {
+			thrown = ex;
+		} finally {
+			cleanupRequests();
+			dispatcher.close(thrown == null ? new CancellationException() : thrown);
+		}
+		LOG.debug("loop terminated");
+	}
+
+	private void loop() throws Exception {
+		while (isActive()) {
+			try {
+				dispatcher.dispatch(deque.take());
+			} catch (InterruptedException e) {
+				if (isActive()) {
+					LOG.debug("interrupted while waiting for a request (continue waiting)", e);
+				} else {
+					Thread.currentThread().interrupt();
+				}
+			}
+		}
+	}
+
+	private void cleanupRequests() {
+		Throwable cause = thrown == null ? new CancellationException() : thrown;
+		List<ChannelStateWriteRequest> drained = new ArrayList<>();
+		deque.drainTo(drained);
+		LOG.info("discarding {} drained requests", drained.size());
+		for (ChannelStateWriteRequest request : drained) {
+			request.cancel(cause);
+		}
+	}
+
+	@Override
+	public void start() throws IllegalStateException {
+		this.thread.start();
+	}
+
+	@Override
+	public void submit(ChannelStateWriteRequest request) throws Exception {
+		submitInternal(request, () -> deque.add(request));
+	}
+
+	@Override
+	public void submitPriority(ChannelStateWriteRequest request) throws Exception {
+		submitInternal(request, () -> deque.addFirst(request));
+	}
+
+	private void submitInternal(ChannelStateWriteRequest request, RunnableWithException action) throws Exception {
+		try {
+			action.run();
+		} catch (Exception ex) {
+			request.cancel(ex);
+			throw ex;
+		}
+		ensureRunning();
+	}
+
+	private void ensureRunning() throws Exception {
+		// this check should be performed *at least after* enqueuing a request
+		// checking before is not enough because (check + enqueue) is not atomic
+		if (!isActive()) {
+			cleanupRequests();
+			throw ExceptionUtils.firstOrSuppressed(new IllegalStateException("not running"), thrown);
+		}
+	}
+
+	private boolean isActive() {
+		return !wasClosed && thread.isAlive();
+	}
+
+	@Override
+	public void close() throws IOException {
+		wasClosed = true;
+		while (thread.isAlive()) {
+			thread.interrupt();
+			try {
+				thread.join();
+			} catch (InterruptedException e) {
+				if (!thread.isAlive()) {
```

Comment:
the condition should be `thread.isAlive`, not `!thread.isAlive()`?

Reply:
This will cause `join` to exit immediately (because the interrupt flag is set) and the loop will be essentially busy waiting.